### ISSUE TITLE
C++ ONNX Inference Gives Different Results Compared to Python for Ultralytics YOLOv11 Classify Model

### ISSUE THREAD

**Issue by 154775258** [ultralytics/ultralytics#23054]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Description
I am deploying the Ultralytics YOLOv11 classification model using C++ and ONNX Runtime. However, I am getting inconsistent inference results between the C++ implementation and the original Python version.

The preprocessing step in C++ is performed using the following function, which resizes and normalizes the input image before feeding it to the ONNX model:

cpp
std::vector<float> Yolo_classify::resize_and_normalization(cv::Mat input_img, int target_size) {
    cv::Mat nchw;
    cv::dnn::blobFromImage(input_img, nchw, 1 / 255.f, cv::Size(target_size, target_size), cv::Scalar(0, 0, 0), true, false, CV_32F);

    // Prepare output vector
    std::vector<float> output;
    output.resize(3 * target_size * target_size);

    // Store data in CHW order
    const float* channel_data = nchw.ptr<float>();
    int end = 3 * target_size * target_size;
    for (int i = 0; i < end; ++i) {
        output[i] = channel_data[i];
    }

    return output;
}
Issue
The outputs from the C++ preprocessing and inference do not match the Python results, even when using the same input image and model. I suspect there may be an issue with the normalization or channel ordering in the C++ preprocessing step.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @154775258, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

This is an automated response 🤖—an Ultralytics engineer will also assist soon ✅.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. In your case (Python vs C++/ONNX Runtime mismatch), an MRE is especially helpful and should include:
-Exact model used (e.g., YOLO11 classify weights) and how it was exported to ONNX
-A single test image (or instructions to reproduce with a public/placeholder image)
-The exact Python inference snippet and outputs
-The exact C++ inference snippet (including ONNX Runtime session options/input tensor shape) and outputs
-Your full environment details (OS, compiler, OpenCV version, ONNX Runtime version, `ultralytics` version)

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
You need LetterBox preprocessing:

https://github.com/ultralytics/ultralytics/blob/794f9f1c5f35b1b4329b1886076ce0ed4b45a96d/examples/YOLOv8-LibTorch-CPP-Inference/main.cc#L27

**Comment by 154775258** [comment]
> You need LetterBox preprocessing:
> 
> [ultralytics/examples/YOLOv8-LibTorch-CPP-Inference/main.cc](https://github.com/ultralytics/ultralytics/blob/794f9f1c5f35b1b4329b1886076ce0ed4b45a96d/examples/YOLOv8-LibTorch-CPP-Inference/main.cc#L27)
> 
> Line 27 in [794f9f1](/ultralytics/ultralytics/commit/794f9f1c5f35b1b4329b1886076ce0ed4b45a96d)
> 
>  float letterbox(cv::Mat &input_image, cv::Mat &output_image, const std::vector<int> &target_size) {

I have updated my preprocessing code to include letterbox resizing as follows:

cpp
std::vector<float> Yolo_classify::resize_and_normalization(cv::Mat input_img, int target_size) {
    // Use letterbox for resize
    cv::Mat resized_img;
    std::vector<int> target_dims = {target_size, target_size};
    letterbox(input_img, resized_img, target_dims);
    
    // Use blobFromImage for normalization and CHW conversion
    cv::Mat nchw;
    cv::dnn::blobFromImage(resized_img, nchw, 1 / 255.f, cv::Size(), cv::Scalar(0, 0, 0), true, false, CV_32F);

    // Prepare output vector
    std::vector<float> output;
    output.resize(3 * target_size * target_size);

    // Store data in CHW order
    const float* channel_data = nchw.ptr<float>();
    int end = 3 * target_size * target_size;
    for (int i = 0; i < end; ++i) {
        output[i] = channel_data[i];
    }

    return output;
}
but the output results of C++ and Python still exhibit inconsistencies.

**Comment by Y-T-G** [comment]
PyTorch runs with minimum rectangle:

https://github.com/ultralytics/ultralytics/issues/16388#issuecomment-2363289302

You can test Ultralytics inference with `rect=False` and compare

**Comment by 154775258** [comment]
> rect=False

i use rect=false in pytorch 
the predict is image 1/1 G:\DeepLearn\VIS_WEB\data\test\154775258\image_test\317(1644)_2025_12_2615_07_12.jpg: 32x32 2 0.69, 1 0.30, 5 0.00, 6 0.00, 7 0.00, 5.4ms
c++ predict is E:\desktop\pic\317 (1644).jpg  class: 2  Conf: 0.5018

**Comment by Y-T-G** [comment]
For classification, it actually doesn't use letterbox. It uses CenterCrop.

https://github.com/ultralytics/ultralytics/blob/e2724c45f4f9b3ffa28a68b62b1fb1d16fd9191f/ultralytics/data/augment.py#L2447

**Comment by 154775258** [comment]
> For classification, it actually doesn't use letterbox. It uses CenterCrop.
> 
> [ultralytics/ultralytics/data/augment.py](https://github.com/ultralytics/ultralytics/blob/e2724c45f4f9b3ffa28a68b62b1fb1d16fd9191f/ultralytics/data/augment.py#L2447)
> 
> Line 2447 in [e2724c4](/ultralytics/ultralytics/commit/e2724c45f4f9b3ffa28a68b62b1fb1d16fd9191f)
> 
>  def classify_transforms(

Thank you for your help. The issue has been resolved.

**Comment by glenn-jocher** [comment]
Glad you got it resolved—if you can, please paste the final C++ preprocessing you used (matching `Resize` → `CenterCrop` → `Normalize` as in [`classify_transforms`](https://docs.ultralytics.com/reference/data/augment/)) so others can reproduce, and we’ll close this out.

**Comment by 154775258** [comment]
> Glad you got it resolved—if you can, please paste the final C++ preprocessing you used (matching `Resize` → `CenterCrop` → `Normalize` as in [`classify_transforms`](https://docs.ultralytics.com/reference/data/augment/)) so others can reproduce, and we’ll close this out.

this is my code
`class ClassifyTransforms {
private:
    cv::Size target_size_;
    std::vector<float> mean_;
    std::vector<float> std_;
    int interpolation_;

public:
    // 构造函数
    ClassifyTransforms(
        cv::Size size = cv::Size(224, 224),
        std::vector<float> mean = { 0, 0, 0 },  // 常见的ImageNet归一化参数
        std::vector<float> std = { 1, 1, 1 },
        std::string interpolation = "BILINEAR"
    ) : target_size_(size), mean_(mean), std_(std) {
        // 设置插值方法
        if (interpolation == "NEAREST") {
            interpolation_ = cv::INTER_NEAREST;
        }
        else if (interpolation == "BILINEAR") {
            interpolation_ = cv::INTER_LINEAR;
        }
        else if (interpolation == "BICUBIC") {
            interpolation_ = cv::INTER_CUBIC;
        }
        else {
            throw std::invalid_argument("Interpolation must be 'NEAREST', 'BILINEAR' or 'BICUBIC'");
        }

        if (mean_.size() != 3 || std_.size() != 3) {
            throw std::invalid_argument("Mean and std must have 3 values for RGB channels");
        }
    }

    // 重载构造函数，接受整数作为正方形尺寸
    ClassifyTransforms(
        int size = 224,
        std::vector<float> mean = { 0, 0, 0 },
        std::vector<float> std = { 1, 1, 1 },
        std::string interpolation = "BILINEAR"
    ) : ClassifyTransforms(cv::Size(size, size), mean, std, interpolation) {
    }

    // 主处理函数
    cv::Mat operator()(const cv::Mat& input_image) {
        if (input_image.empty()) {
            throw std::invalid_argument("Input image is empty");
        }

        cv::Mat processed = input_image.clone();

        // 步骤1: 调整大小
        cv::Mat resized;
        if (target_size_.width == target_size_.height) {
            // 正方形目标尺寸：保持宽高比，调整最短边
            int short_side = std::min(processed.cols, processed.rows);
            float scale = static_cast<float>(target_size_.width) / short_side;
            cv::Size new_size(
                static_cast<int>(processed.cols * scale),
                static_cast<int>(processed.rows * scale)
            );
            cv::resize(processed, resized, new_size, 0, 0, interpolation_);
        }
        else {
            // 非正方形目标尺寸：直接调整到目标尺寸
            cv::resize(processed, resized, target_size_, 0, 0, interpolation_);
        }

        // 步骤2: 中心裁剪
        cv::Mat cropped;
        int crop_x = std::max(0, (resized.cols - target_size_.width) / 2);
        int crop_y = std::max(0, (resized.rows - target_size_.height) / 2);
        cv::Rect roi(
            crop_x,
            crop_y,
            std::min(target_size_.width, resized.cols - crop_x),
            std::min(target_size_.height, resized.rows - crop_y)
        );
        cropped = resized(roi).clone();

        // 确保裁剪后的图像尺寸正确
        if (cropped.cols != target_size_.width || cropped.rows != target_size_.height) {
            cv::resize(cropped, cropped, target_size_, 0, 0, interpolation_);
        }

        // 步骤3: 转换为浮点数并归一化到[0, 1]
        cv::Mat float_image;
        if (cropped.depth() == CV_8U) {
            cropped.convertTo(float_image, CV_32F, 1.0 / 255.0);
        }
        else if (cropped.depth() == CV_32F) {
            float_image = cropped.clone();
        }
        else {
            cropped.convertTo(float_image, CV_32F);
        }

        // 步骤4: 将BGR转换为RGB（OpenCV默认是BGR）
        cv::Mat rgb_image;
        cv::cvtColor(float_image, rgb_image, cv::COLOR_BGR2RGB);

        // 步骤5: 分离通道并归一化
        std::vector<cv::Mat> channels(3);
        cv::split(rgb_image, channels);

        // 对每个通道应用归一化：(channel - mean) / std
        for (int i = 0; i < 3; i++) {
            channels[i] = (channels[i] - mean_[i]) / std_[i];
        }

        // 步骤6: 合并通道
        cv::Mat normalized;
        cv::merge(channels, normalized);

        // 步骤7: 转换为CHW格式（如果需要，虽然OpenCV通常是HWC）
        // 注意：很多深度学习框架使用CHW，但这里我们保持OpenCV的HWC格式
        // 如果需要CHW，可以在这里添加转置操作

        cv::Mat chw;

        // 方法1: 使用split和reshape（更高效）
        cv::split(normalized, channels); // 再次分离通道
        chw.create(3 * target_size_.height * target_size_.width, 1, CV_32F);

        // 将每个通道的数据复制到CHW格式的矩阵中
        for (int i = 0; i < 3; i++) {
            cv::Mat channel_flat = channels[i].reshape(1, target_size_.height * target_size_.width);
            channel_flat.copyTo(chw.rowRange(i * target_size_.height * target_size_.width,
                (i + 1) * target_size_.height * target_size_.width));
        }

        // 将一维向量reshape为3xHxW的矩阵
        chw = chw.reshape(1, { 3, target_size_.height, target_size_.width });

        return chw;
    }

    // 获取预处理后的图像作为一维向量（用于输入到某些模型）
    std::vector<float> getVector(const cv::Mat& input_image) {
        cv::Mat processed = this->operator()(input_image);

        // 将OpenCV Mat转换为一维向量
        std::vector<float> result;
        if (processed.isContinuous()) {
            result.assign((float*)processed.data,
                (float*)processed.data + processed.total() * processed.channels());
        }
        else {
            // 如果不是连续的，需要逐行复制
            for (int i = 0; i < processed.rows; i++) {
                const float* row = processed.ptr<float>(i);
                result.insert(result.end(), row, row + processed.cols * processed.channels());
            }
        }

        return result;
    }

    // 获取目标尺寸
    cv::Size getTargetSize() const { return target_size_; }

    // 获取均值
    std::vector<float> getMean() const { return mean_; }

    // 获取标准差
    std::vector<float> getStd() const { return std_; }
};

std::vector<float> Yolo_classify::resize_and_normalization(cv::Mat input_img, int target_size) {
    ClassifyTransforms transform(target_size);
    return transform.getVector(input_img);
}

`

**Comment by glenn-jocher** [comment]
Thanks for sharing—this matches Ultralytics’ classification pipeline (`Resize` shortest-edge → `CenterCrop` → `ToTensor` → `Normalize`) from [`classify_transforms`](https://docs.ultralytics.com/reference/data/augment/); the main thing to double-check is that your `mean_`/`std_` match the exact `DEFAULT_MEAN`/`DEFAULT_STD` used by Ultralytics (your current defaults `{0,0,0}` / `{1,1,1}` will shift outputs). You can print the expected values from your installed version with:

```python
from ultralytics.data.augment import DEFAULT_MEAN, DEFAULT_STD
print(DEFAULT_MEAN, DEFAULT_STD)
```

**Comment by 154775258** [comment]
> Thanks for sharing—this matches Ultralytics’ classification pipeline (`Resize` shortest-edge → `CenterCrop` → `ToTensor` → `Normalize`) from [`classify_transforms`](https://docs.ultralytics.com/reference/data/augment/); the main thing to double-check is that your `mean_`/`std_` match the exact `DEFAULT_MEAN`/`DEFAULT_STD` used by Ultralytics (your current defaults `{0,0,0}` / `{1,1,1}` will shift outputs). You can print the expected values from your installed version with:
> 
> from ultralytics.data.augment import DEFAULT_MEAN, DEFAULT_STD
> print(DEFAULT_MEAN, DEFAULT_STD)

this code output is (0.0, 0.0, 0.0) (1.0, 1.0, 1.0)。

**Comment by glenn-jocher** [comment]
Good to hear—if `DEFAULT_MEAN/DEFAULT_STD` are `(0,0,0)` / `(1,1,1)` in your install, then normalization isn’t the source of any remaining mismatch; the next most common culprit is an off-by-1 resize rounding difference (which changes the `CenterCrop` window), so make sure your “shortest-edge resize” uses the same rounding policy as torchvision (avoid implicit `int` truncation), and that you’re matching the exact RGB conversion flow used in [`ClassificationPredictor.preprocess`](https://docs.ultralytics.com/reference/models/yolo/classify/predict/) before feeding `NCHW` float input to ONNX Runtime.

```python
import cv2
from PIL import Image
from ultralytics.data.augment import classify_transforms

im = cv2.imread("path/to.jpg")  # BGR
t = classify_transforms(size=32)(Image.fromarray(cv2.cvtColor(im, cv2.COLOR_BGR2RGB)))  # set size=your imgsz
print(t.shape, t.flatten()[:20].tolist())  # compare these floats to C++ to pinpoint where drift starts
```
